src/encoding.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * encoding.js
+ *
+ * This NPM module provide encoding and decoding utilities for interacting with
+ * dFusion smart contracts where manual byte encoding was needed because of
+ * solidity ABI limitations.
+ */
+import BN from "bn.js";
+import assert from "assert";
+
+export interface Order<T = string> {
+  user: string;
+  sellTokenBalance: T;
+  buyToken: number;
+  sellToken: number;
+  validFrom: number;
+  validUntil: number;
+  priceNumerator: T;
+  priceDenominator: T;
+  remainingAmount: T;
+}
+
+export interface IndexedOrder<T> extends Order<T> {
+  orderId: number;
+}
+
+class OrderBuffer {
+  index = 2; // skip '0x'
+  constructor(public bytes: string) {}
+
+  readBytes = (size: number) =>
+    this.bytes.slice(this.index, (this.index += size * 2));
+
+  decodeAddr = () => `0x${this.readBytes(20)}`;
+  decodeInt = (size: number): BN => new BN(this.readBytes(size / 8), 16);
+  decodeNumber = (size: number) => parseInt(this.readBytes(size / 8), 16);
+}
+
+function decodeOrder(bytes: OrderBuffer): Order<BN> {
+  return {
+    user: bytes.decodeAddr(),
+    sellTokenBalance: bytes.decodeInt(256),
+    buyToken: bytes.decodeNumber(16),
+    sellToken: bytes.decodeNumber(16),
+    validFrom: bytes.decodeNumber(32),
+    validUntil: bytes.decodeNumber(32),
+    priceNumerator: bytes.decodeInt(128),
+    priceDenominator: bytes.decodeInt(128),
+    remainingAmount: bytes.decodeInt(128),
+  };
+}
+
+function decodeIndexedOrder(bytes: OrderBuffer): IndexedOrder<BN> {
+  return {
+    ...decodeOrder(bytes),
+    orderId: bytes.decodeNumber(16),
+  };
+}
+
+function decodeOrdersInternal<T>(
+  bytes: string,
+  decodeFunction: (x: OrderBuffer) => T,
+  width: number,
+): T[] {
+  if (bytes === null || bytes === undefined || bytes.length === 0) {
```

Comment:
I was surprised at first that TS allows a comparison between `null`or `undefined` and a string, but this is indeed a [language feature](https://www.typescriptlang.org/docs/handbook/basic-types.html#null-and-undefined). I assume we don't want to use `strictNullChecks`?

Reply:
We are using strict mode which includes strict null checks (https://github.com/gnosis/dex-contracts/blob/4a3e99e1e5746bfb4f7ed7db304d86c3094681b9/tsconfig.json#L26-L27)

Reply:
But then why isn't TS failing to compile `bytes === null` and `bytes === undefined`? They should always be false, since `bytes` is declared as `bytes: string`, it'd be like adding `bytes === 5` to that line.
The best theory I came up with is that `bytes` is considered an [optional parameters](https://www.typescriptlang.org/v2/docs/handbook/type-checking-javascript-files.html#function-parameters-are-optional-by-default) (but this not a great explanation since this is a pre-ES2015 behavior and we specify `"target": "es2015"`) and both `null` and `undefined` could be used to represent an unused parameter. This goes in addition to the fact that some operators [gobble `null` and `undefined`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#expression-operators), so it might be normally not noticed.

Reply:
Good question

https://github.com/microsoft/TypeScript/issues/11920#issuecomment-256961196

> We intentionally allow this for the sake of defensive programming (i.e. defending against missing inputs from non-TS code). If there's enough demand we could add a flag or something.